File: src/test_core/fmg/core/mosaic/MosaicInitDataTest.java
package fmg.core.mosaic;

import static org.junit.Assert.assertEquals;

import java.util.List;

import org.junit.*;

import fmg.common.Logger;
import fmg.common.notifier.PropertyChangeExecutor;
import fmg.core.types.EMosaic;
import fmg.core.types.EMosaicGroup;
import fmg.core.types.ESkillLevel;
import io.reactivex.Flowable;

public class MosaicInitDataTest {

    @BeforeClass
    public static void setup() {
        Logger.info(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");
        Logger.info("> MosaicInitDataTest::setup");

        MosaicModelTest.ProjSettings();

        Flowable.just("UI factory inited...").subscribe(Logger::info);
    }

    @Before
    public void before() {
        Logger.info("======================================================");
    }

    @AfterClass
    public static void after() {
        Logger.info("======================================================");
        Logger.info("< MosaicInitDataTest closed");
        Logger.info("<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<");
    }

    private static MosaicInitData createMosaicInitData() {
        MosaicInitData initData = new MosaicInitData();
        assertEquals(MosaicInitData.DEFAULT_MOSAIC_TYPE , initData.getMosaicType());
        assertEquals(MosaicInitData.DEFAULT_SIZE_FIELD_M, initData.getSizeField().m);
        assertEquals(MosaicInitData.DEFAULT_SIZE_FIELD_N, initData.getSizeField().n);
        assertEquals(MosaicInitData.DEFAULT_COUNT_MINES , initData.getCountMines());
        assertEquals(MosaicInitData.DEFAULT_SKILL_LEVEL , initData.getSkillLevel());

        assertEquals(MosaicInitData.DEFAULT_MOSAIC_TYPE.getGroup(), initData.getMosaicGroup());
        return initData;
    }

    @Test
    public void checkTheImpossibilitySetCustomSkillLevelTest() {
        Logger.info("> MosaicInitDataTest::checkTheImpossibilitySetCustomSkillLevelTest");
        try {
            new PropertyChangeExecutor<>(MosaicInitDataTest::createMosaicInitData).run(10, 1000,
               initData -> {
                   initData.setSkillLevel(ESkillLevel.eCustom);
                   Assert.fail();
               }, (initData, modifiedProperties) -> {});
        } catch (Exception ex) {
            assertEquals(IllegalArgumentException.class, ex.getClass());
        }
    }

    @Test
    public void checkIfMosaicTypeIsChangedThenCountMinesWillAlsoBeChangedTest() {
        Logger.info("> MosaicInitDataTest::checkIfMosaicTypeIsChangedThenCountMinesWillAlsoBeChangedTest");
        new PropertyChangeExecutor<>(MosaicInitDataTest::createMosaicInitData).run(300, 5000,
            initData -> {
                initData.setMosaicType(EMosaic.eMosaicRhombus1);
            }, (initData, modifiedProperties) -> {
                Assert.assertTrue  (   modifiedProperties.containsKey(MosaicInitData.PROPERTY_MOSAIC_TYPE));
                Assert.assertEquals(1, modifiedProperties.get(        MosaicInitData.PROPERTY_MOSAIC_TYPE).first.intValue());
                Assert.assertTrue  (   modifiedProperties.containsKey(MosaicInitData.PROPERTY_COUNT_MINES));
                Assert.assertEquals(1, modifiedProperties.get(        MosaicInitData.PROPERTY_COUNT_MINES).first.intValue());
                Assert.assertEquals(2, modifiedProperties.size());
                Assert.assertEquals(EMosaic.eMosaicRhombus1, initData.getMosaicType());
                Assert.assertEquals(initData.getSkillLevel().getNumberMines(EMosaic.eMosaicRhombus1), initData.getCountMines());
            });
    }

    @Test
    public void checkNoRepeatNotificationsTest() {
        Logger.info("> MosaicInitDataTest::checkNoRepeatNotificationsTest");
            new PropertyChangeExecutor<>(MosaicInitDataTest::createMosaicInitData).run(300, 5000,
                initData -> {
                    Logger.info("    initData.countMines={0}", initData.getCountMines());
                    initData.setMosaicType(EMosaic.eMosaicRhombus1);
                    Logger.info("    initData.countMines={0}", initData.getCountMines());
                    initData.setMosaicType(EMosaic.eMosaicHexagon1);
                    Logger.info("    initData.countMines={0}", initData.getCountMines());
                }, (initData, modifiedProperties) -> {
                    Assert.assertTrue  (   modifiedProperties.containsKey(MosaicInitData.PROPERTY_MOSAIC_TYPE));
                    Assert.assertEquals(1, modifiedProperties.get(        MosaicInitData.PROPERTY_MOSAIC_TYPE).first.intValue());
                    Assert.assertTrue  (   modifiedProperties.containsKey(MosaicInitData.PROPERTY_MOSAIC_GROUP));
                    Assert.assertEquals(1, modifiedProperties.get(        MosaicInitData.PROPERTY_MOSAIC_GROUP).first.intValue());
                    Assert.assertTrue  (  !modifiedProperties.containsKey(MosaicInitData.PROPERTY_COUNT_MINES));
                    Assert.assertEquals(2, modifiedProperties.size());
                    Assert.assertEquals(EMosaic.eMosaicHexagon1, initData.getMosaicType());
                    Assert.assertEquals(initData.getSkillLevel().getNumberMines(EMosaic.eMosaicHexagon1), initData.getCountMines());
                });
    }

    @Test
    public void checkChangedMosaicGroupTest() {
        Logger.info("> MosaicInitDataTest::checkChangedMosaicGroupTest");
        new PropertyChangeExecutor<>(MosaicInitDataTest::createMosaicInitData).run(300, 5000,
            initData -> {
                initData.setMosaicType(EMosaic.eMosaicHexagon1);
            }, (initData, modifiedProperties) -> {
                Assert.assertTrue(modifiedProperties.containsKey(MosaicInitData.PROPERTY_MOSAIC_GROUP));
            });
    }

    @Test
    public void checkNoChangedMosaicGroupTest() {
        Logger.info("> MosaicInitDataTest::checkNoChangedMosaicGroupTest");
        new PropertyChangeExecutor<>(MosaicInitDataTest::createMosaicInitData).run(300, 5000,
            initData -> {
                initData.setMosaicType(EMosaic.eMosaicRhombus1);
            }, (initData, modifiedProperties) -> {
                Assert.assertFalse(modifiedProperties.containsKey(MosaicInitData.PROPERTY_MOSAIC_GROUP));
            });
    }

    @Test
    public void checkRestoreIndexInGroupTest() {
        Logger.info("> MosaicInitDataTest::checkRestoreIndexInGroupTest");

        new PropertyChangeExecutor<>(MosaicInitDataTest::createMosaicInitData).run(10, 1000,
            initData -> {
                final int checkOrdinal = 3;

                // 1. select another mosaic in current group
                List<EMosaic> mosaicsInOldGroup = initData.getMosaicGroup().getMosaics();
                initData.setMosaicType(mosaicsInOldGroup.get(checkOrdinal));

                // 2. change group
                initData.setMosaicGroup(EMosaicGroup.eTriangles);

                // 3. check ordinal in new group
                Assert.assertEquals(checkOrdinal, initData.getMosaicType().getOrdinalInGroup());
            }, (initData, modifiedProperties) -> {});
    }
}

File: src/test_core/fmg/core/mosaic/MosaicInitDataTest.cs
using System;
using System.Linq;
using System.Threading.Tasks;
using fmg.common;
using fmg.common.notifier;
using fmg.core.types;

namespace fmg.core.mosaic {

    public abstract class MosaicInitDataTest {

        protected abstract void AssertEqual(int    expected, int    actual);
        protected abstract void AssertEqual(object expected, object actual);
        protected abstract void AssertTrue(bool condition);
        protected abstract void AssertFalse(bool condition);
        protected abstract void AssertFail();

        public virtual void Setup() {
            LoggerSimple.Put(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");
            LoggerSimple.Put("> " + nameof(MosaicInitDataTest) + "::" + nameof(Setup));
        }

        public virtual void Before() {
            LoggerSimple.Put("======================================================");
        }

        public virtual void After() {
            LoggerSimple.Put("======================================================");
            LoggerSimple.Put("< " + nameof(MosaicInitDataTest) + " closed");
            LoggerSimple.Put("<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<");
        }

        private MosaicInitData CreateMosaicInitData() {
            var initData = new MosaicInitData();
            AssertEqual(MosaicInitData.DEFAULT_MOSAIC_TYPE , initData.MosaicType);
            AssertEqual(MosaicInitData.DEFAULT_SIZE_FIELD_M, initData.SizeField.m);
            AssertEqual(MosaicInitData.DEFAULT_SIZE_FIELD_N, initData.SizeField.n);
            AssertEqual(MosaicInitData.DEFAULT_MINES_COUNT , initData.MinesCount);
            AssertEqual(MosaicInitData.DEFAULT_SKILL_LEVEL , initData.SkillLevel);

            AssertEqual(MosaicInitData.DEFAULT_MOSAIC_TYPE.GetGroup(), initData.MosaicGroup);
            return initData;
        }


        public virtual void CheckTheImpossibilitySetCustomSkillLevelTest() {
            LoggerSimple.Put("> " + nameof(MosaicInitDataTest) + "::" + nameof(CheckTheImpossibilitySetCustomSkillLevelTest));
            using (var initData = CreateMosaicInitData()) try {
                initData.SkillLevel = ESkillLevel.eCustom;
                AssertFail();
            } catch (Exception ex) {
                AssertEqual(typeof(ArgumentException), ex.GetType());
            }
        }

        public virtual async Task CheckIfMosaicTypeIsChangedThenMinesCountWillAlsoBeChangedTest() {
            LoggerSimple.Put("> " + nameof(MosaicInitDataTest) + "::" + nameof(CheckIfMosaicTypeIsChangedThenMinesCountWillAlsoBeChangedTest));

            await new PropertyChangeExecutor<MosaicInitData>(() => CreateMosaicInitData()).Run(100, 1000,
                initData => {
                    initData.MosaicType = EMosaic.eMosaicRhombus1;
                }, (initData, modifiedProperties) => {
                    AssertTrue  (  modifiedProperties.ContainsKey(nameof(MosaicInitData.MosaicType)));
                    AssertEqual(1, modifiedProperties[            nameof(MosaicInitData.MosaicType)]);
                    AssertTrue  (  modifiedProperties.ContainsKey(nameof(MosaicInitData.MinesCount)));
                    AssertEqual(1, modifiedProperties[            nameof(MosaicInitData.MinesCount)]);
                    AssertEqual(2, modifiedProperties.Count);
                    AssertEqual(EMosaic.eMosaicRhombus1, initData.MosaicType);
                    AssertEqual(initData.SkillLevel.GetNumberMines(EMosaic.eMosaicRhombus1), initData.MinesCount);
                });
        }

        public virtual async Task CheckNoRepeatNotificationsTest() {
            LoggerSimple.Put("> " + nameof(MosaicInitDataTest) + "::" + nameof(CheckNoRepeatNotificationsTest));

            await new PropertyChangeExecutor<MosaicInitData>(() => CreateMosaicInitData()).Run(100, 1000,
                initData => {
                    LoggerSimple.Put("    initData.minesCount={0}", initData.MinesCount);
                    initData.MosaicType = EMosaic.eMosaicRhombus1;
                    LoggerSimple.Put("    initData.minesCount={0}", initData.MinesCount);
                    initData.MosaicType = EMosaic.eMosaicHexagon1;
                    LoggerSimple.Put("    initData.minesCount={0}", initData.MinesCount);
                }, (initData, modifiedProperties) => {
                    AssertTrue  (  modifiedProperties.ContainsKey(nameof(MosaicInitData.MosaicType)));
                    AssertEqual(1, modifiedProperties[            nameof(MosaicInitData.MosaicType)]);
                    AssertTrue  (  modifiedProperties.ContainsKey(nameof(MosaicInitData.MosaicGroup)));
                    AssertEqual(1, modifiedProperties[            nameof(MosaicInitData.MosaicGroup)]);
                    AssertTrue  ( !modifiedProperties.ContainsKey(nameof(MosaicInitData.MinesCount)));
                    AssertEqual(2, modifiedProperties.Count);
                    AssertEqual(EMosaic.eMosaicHexagon1, initData.MosaicType);
                    AssertEqual(initData.SkillLevel.GetNumberMines(EMosaic.eMosaicHexagon1), initData.MinesCount);
                });
        }

        public virtual async Task CheckChangedMosaicGroupTest() {
            LoggerSimple.Put("> " + nameof(MosaicInitDataTest) + "::" + nameof(CheckChangedMosaicGroupTest));

            await new PropertyChangeExecutor<MosaicInitData>(() => CreateMosaicInitData()).Run(100, 1000,
                initData => {
                    initData.MosaicType = EMosaic.eMosaicHexagon1;
                }, (initData, modifiedProperties) => {
                    AssertTrue(modifiedProperties.ContainsKey(nameof(MosaicInitData.MosaicGroup)));
                });
        }

        public virtual async Task CheckNoChangedMosaicGroupTest() {
            LoggerSimple.Put("> " + nameof(MosaicInitDataTest) + "::" + nameof(CheckNoChangedMosaicGroupTest));

            await new PropertyChangeExecutor<MosaicInitData>(() => CreateMosaicInitData()).Run(100, 1000,
                initData => {
                    initData.MosaicType = EMosaic.eMosaicRhombus1;
                }, (initData, modifiedProperties) => {
                    AssertFalse(modifiedProperties.ContainsKey(nameof(MosaicInitData.MosaicGroup)));
                });
        }

        public virtual void CheckRestoreIndexInGroupTest() {
            LoggerSimple.Put("> " + nameof(MosaicInitDataTest) + "::" + nameof(CheckRestoreIndexInGroupTest));

            using (var initData = CreateMosaicInitData()) {
                const int checkOrdinal = 3;

                // 1. select another mosaic in current group
                var mosaicsInOldGroup = initData.MosaicGroup.GetMosaics();
                initData.MosaicType = mosaicsInOldGroup[checkOrdinal];

                // 2. change group
                initData.MosaicGroup = EMosaicGroup.eTriangles;

                // 3. check ordinal in new group
                AssertEqual(checkOrdinal, initData.MosaicType.GetOrdinalInGroup());
            }
        }

    }

}
